c10/core/TensorImpl.h

```diff
@@ -826,6 +826,11 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
    */
   void set_autograd_meta(std::unique_ptr<c10::AutogradMetaInterface> autograd_meta) {
     autograd_meta_ = std::move(autograd_meta);
+    if (autograd_meta_) {
```

Comment:
this seems like the wrong layer of abstraction to do this?  Meaning, you are keeping VariableTensorId in sync with autograd_meta (which is reasonable), but I think these are all called by more-semantically-meaningful functions like "make_variable" or similar (and it seems fine to make this an error to be called from any other place).  I could be wrong and this would make it overly complicated, it depends on the implementation complexity.

Reply:
Sort of. I agree with your general sentiment here, but at the end of the day, `autograd_meta_` and `type_set_` are private members on TensorImpl, and so methods on TensorImpl are the best place to ensure that the invariant is upheld, since it's easy to audit all of the sites which access these fields. It's probably reasonable to try to enforce these invariants at a higher level, but my preference would be that we'd start migrating some variable functionality into TensorImpl itself.